In [1]:

import numpy as np
import pandas as pd


In [2]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [4]:
df = pd.read_csv('StudentPerformanceFactors.csv')

In [ ]:
df.head()

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
0             23          84                  Low                High   
1             19          64                  Low              Medium   
2             24          98               Medium              Medium   
3             29          89                  Low              Medium   
4             19          92               Medium              Medium   

  Extracurricular_Activities  Sleep_Hours  Previous_Scores Motivation_Level  \
0                         No            7               73              Low   
1                         No            8               59              Low   
2                        Yes            7               91           Medium   
3                        Yes            8               98           Medium   
4                        Yes            6               65           Medium   

  Internet_Access  Tutoring_Sessions Family_Income Teacher_Quality  \
0             Yes                  0           Low          Medium   
1             Yes                  2        Medium          Medium   
2             Yes                  2        Medium          Medium   
3             Yes                  1        Medium          Medium   
4             Yes                  3        Medium            High   

  School_Type Peer_Influence  Physical_Activity Learning_Disabilities  \
0      Public       Positive                  3                    No   
1      Public       Negative                  4                    No   
2      Public        Neutral                  4                    No   
3      Public       Negative                  4                    No   
4      Public        Neutral                  4                    No   

  Parental_Education_Level Distance_from_Home  Gender  Exam_Score  
0              High School               Near    Male          67  
1                  College           Moderate  Female          61  
2             Postgraduate               Near    Male          74  
3              High School           Moderate    Male          71  
4                  College               Near  Female          70

In [5]:
df.dtypes

Hours_Studied                  int64
Attendance                     int64
Parental_Involvement          object
Access_to_Resources           object
Extracurricular_Activities    object
Sleep_Hours                    int64
Previous_Scores                int64
Motivation_Level              object
Internet_Access               object
Tutoring_Sessions              int64
Family_Income                 object
Teacher_Quality               object
School_Type                   object
Peer_Influence                object
Physical_Activity              int64
Learning_Disabilities         object
Parental_Education_Level      object
Distance_from_Home            object
Gender                        object
Exam_Score                     int64
dtype: object

In [6]:
categorical_columns = df.select_dtypes(include=['object']).columns

In [7]:
df_encoded = pd.get_dummies(df, columns=categorical_columns)

In [8]:
df_encoded.head()

Hours_Studied  Attendance  Sleep_Hours  Previous_Scores  Tutoring_Sessions  \
0             23          84            7               73                  0   
1             19          64            8               59                  2   
2             24          98            7               91                  2   
3             29          89            8               98                  1   
4             19          92            6               65                  3   

   Physical_Activity  Exam_Score  Parental_Involvement_High  \
0                  3          67                      False   
1                  4          61                      False   
2                  4          74                      False   
3                  4          71                      False   
4                  4          70                      False   

   Parental_Involvement_Low  Parental_Involvement_Medium  ...  \
0                      True                        False  ...   
1                      True                        False  ...   
2                     False                         True  ...   
3                      True                        False  ...   
4                     False                         True  ...   

   Learning_Disabilities_No  Learning_Disabilities_Yes  \
0                      True                      False   
1                      True                      False   
2                      True                      False   
3                      True                      False   
4                      True                      False   

   Parental_Education_Level_College  Parental_Education_Level_High School  \
0                             False                                  True   
1                              True                                 False   
2                             False                                 False   
3                             False                                  True   
4                              True                                 False   

   Parental_Education_Level_Postgraduate  Distance_from_Home_Far  \
0                                  False                   False   
1                                  False                   False   
2                                   True                   False   
3                                  False                   False   
4                                  False                   False   

   Distance_from_Home_Moderate  Distance_from_Home_Near  Gender_Female  \
0                        False                     True          False   
1                         True                    False           True   
2                        False                     True          False   
3                         True                    False          False   
4                        False                     True           True   

   Gender_Male  
0         True  
1        False  
2         True  
3         True  
4        False  

[5 rows x 41 columns]

In [9]:
df_encoded = df_encoded.apply(lambda col: col.map({True: 1, False: 0}) if col.dtype == 'bool' else col)

In [10]:
df_encoded.head()

Hours_Studied  Attendance  Sleep_Hours  Previous_Scores  Tutoring_Sessions  \
0             23          84            7               73                  0   
1             19          64            8               59                  2   
2             24          98            7               91                  2   
3             29          89            8               98                  1   
4             19          92            6               65                  3   

   Physical_Activity  Exam_Score  Parental_Involvement_High  \
0                  3          67                          0   
1                  4          61                          0   
2                  4          74                          0   
3                  4          71                          0   
4                  4          70                          0   

   Parental_Involvement_Low  Parental_Involvement_Medium  ...  \
0                         1                            0  ...   
1                         1                            0  ...   
2                         0                            1  ...   
3                         1                            0  ...   
4                         0                            1  ...   

   Learning_Disabilities_No  Learning_Disabilities_Yes  \
0                         1                          0   
1                         1                          0   
2                         1                          0   
3                         1                          0   
4                         1                          0   

   Parental_Education_Level_College  Parental_Education_Level_High School  \
0                                 0                                     1   
1                                 1                                     0   
2                                 0                                     0   
3                                 0                                     1   
4                                 1                                     0   

   Parental_Education_Level_Postgraduate  Distance_from_Home_Far  \
0                                      0                       0   
1                                      0                       0   
2                                      1                       0   
3                                      0                       0   
4                                      0                       0   

   Distance_from_Home_Moderate  Distance_from_Home_Near  Gender_Female  \
0                            0                        1              0   
1                            1                        0              1   
2                            0                        1              0   
3                            1                        0              0   
4                            0                        1              1   

   Gender_Male  
0            1  
1            0  
2            1  
3            1  
4            0  

[5 rows x 41 columns]

In [11]:
df_encoded.dtypes

Hours_Studied                            int64
Attendance                               int64
Sleep_Hours                              int64
Previous_Scores                          int64
Tutoring_Sessions                        int64
Physical_Activity                        int64
Exam_Score                               int64
Parental_Involvement_High                int64
Parental_Involvement_Low                 int64
Parental_Involvement_Medium              int64
Access_to_Resources_High                 int64
Access_to_Resources_Low                  int64
Access_to_Resources_Medium               int64
Extracurricular_Activities_No            int64
Extracurricular_Activities_Yes           int64
Motivation_Level_High                    int64
Motivation_Level_Low                     int64
Motivation_Level_Medium                  int64
Internet_Access_No                       int64
Internet_Access_Yes                      int64
Family_Income_High                       int64
Family_Income_Low                        int64
Family_Income_Medium                     int64
Teacher_Quality_High                     int64
Teacher_Quality_Low                      int64
Teacher_Quality_Medium                   int64
School_Type_Private                      int64
School_Type_Public                       int64
Peer_Influence_Negative                  int64
Peer_Influence_Neutral                   int64
Peer_Influence_Positive                  int64
Learning_Disabilities_No                 int64
Learning_Disabilities_Yes                int64
Parental_Education_Level_College         int64
Parental_Education_Level_High School     int64
Parental_Education_Level_Postgraduate    int64
Distance_from_Home_Far                   int64
Distance_from_Home_Moderate              int64
Distance_from_Home_Near                  int64
Gender_Female                            int64
Gender_Male                              int64
dtype: object

In [12]:
X = df_encoded.drop('Exam_Score', axis=1)
y = df_encoded['Exam_Score']

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [14]:
model = LinearRegression()

In [15]:
model.fit(X_train, y_train)

LinearRegression()

In [16]:
y_pred = model.predict(X_test)

In [17]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [18]:
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)


In [19]:
print(f'Mean Absolute Error: {mae}')
print(f'Mean Squared Error: {mse}')
print(f'R-squared: {r2}')

Mean Absolute Error: 0.44994950022706737
Mean Squared Error: 3.252081227589894
R-squared: 0.7699282240123635


## Model Performance

The performance of the model was evaluated using several metrics:

- **Mean Absolute Error (MAE):** 0.45  
    The average absolute difference between the predicted and actual values. A lower value indicates better performance.

- **Mean Squared Error (MSE):** 3.25  
    The average squared difference between the predicted and actual values. This penalizes larger errors more than MAE.

- **R-squared (R²):** 0.77  
    The coefficient of determination, indicating how well the model explains the variance in the target variable. An R² value closer to 1 suggests a better fit.

These results show that the model performs well, explaining about 77% of the variance in exam scores based on the provided features.
